packages/netlify-cms-lib-util/src/implementation.ts

```diff
@@ -306,3 +325,255 @@ export const runWithLock = async (lock: AsyncLock, func: Function, message: stri
     lock.release();
   }
 };
+
+const LOCAL_KEY = 'git.local';
+
+type LocalTree = {
+  head: string;
+  files: { id: string; name: string; path: string }[];
+};
+
+type GetKeyArgs = {
+  branch: string;
+  folder: string;
+  extension: string;
+  depth: number;
+};
+
+const getLocalKey = ({ branch, folder, extension, depth }: GetKeyArgs) => {
+  return `${LOCAL_KEY}.${branch}.${folder}.${extension}.${depth}`;
+};
+
+type PersistLocalTreeArgs = GetKeyArgs & {
+  localForage: LocalForage;
+  localTree: LocalTree;
+};
+
+type GetLocalTreeArgs = GetKeyArgs & {
+  localForage: LocalForage;
+};
+
+export const persistLocalTree = async ({
+  localForage,
+  localTree,
+  branch,
+  folder,
+  extension,
+  depth,
+}: PersistLocalTreeArgs) => {
+  await localForage.setItem<LocalTree>(
+    getLocalKey({ branch, folder, extension, depth }),
+    localTree,
+  );
+};
+
+export const getLocalTree = async ({
+  localForage,
+  branch,
+  folder,
+  extension,
+  depth,
+}: GetLocalTreeArgs) => {
+  const localTree = await localForage.getItem<LocalTree>(
+    getLocalKey({ branch, folder, extension, depth }),
+  );
+  return localTree;
+};
+
+type GetDiffFromLocalTreeMethods = {
+  getDifferences: (
+    to: string,
+    from: string,
+  ) => Promise<
+    {
+      oldPath: string;
+      newPath: string;
+      status: string;
+      binary: boolean;
+    }[]
+  >;
+  filterFile: (file: { path: string; name: string }) => boolean;
+  getFileId: (path: string) => Promise<string>;
+};
+
+type GetDiffFromLocalTreeArgs = GetDiffFromLocalTreeMethods & {
+  branch: { name: string; sha: string };
+  localTree: LocalTree;
+  folder: string;
+  extension: string;
+  depth: number;
+};
+
+const getDiffFromLocalTree = async ({
```

Comment:
I saw a lot of errors about diffs when working with my crazy erquhart/blank testbed repo - might want to test against it a bit yourself. Ran into some of those random API errors where entries refuse to save, too, but not sure if those are related to this.

Reply:
This function is not used actually for the GitHub backend (only for GitLab and Bitbucket).
The errors you're seeing are probably related to https://github.com/netlify/netlify-cms/pull/3508 as I added more logs to debug https://github.com/netlify/netlify-cms/issues/3506